apps/minifront/src/state/swap/swap-balances-middleware.test.ts

```diff
@@ -0,0 +1,293 @@
+import { beforeEach, describe, expect, it, vi } from 'vitest';
+import { ZQueryState } from '@penumbra-zone/zquery';
+import {
+  Metadata,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import { BalancesResponse } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+import { AddressView } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1/keys_pb';
+import { emptyBalanceResponse } from '../../utils/empty-balance-response';
+import { create } from 'zustand';
+import { AllSlices, initializeStore } from '..';
+
+describe('swapBalancesMiddleware()', () => {
+  const umToken = new Metadata({
+    base: 'upenumbra',
+    display: 'penumbra',
+    denomUnits: [{ denom: 'upenumbra' }, { denom: 'penumbra', exponent: 6 }],
+    symbol: 'UM',
+    penumbraAssetId: { inner: new Uint8Array([1]) },
+  });
+  const usdcToken = new Metadata({
+    base: 'uusdc',
+    display: 'usdc',
+    denomUnits: [{ denom: 'uusdc' }, { denom: 'usdc', exponent: 6 }],
+    symbol: 'USDC',
+    penumbraAssetId: { inner: new Uint8Array([2]) },
+  });
+  const osmoToken = new Metadata({
+    base: 'uosmo',
+    display: 'osmo',
+    denomUnits: [{ denom: 'uosmo' }, { denom: 'osmo', exponent: 6 }],
+    symbol: 'OSMO',
+    penumbraAssetId: { inner: new Uint8Array([3]) },
+  });
+
+  const swappableAssets: ZQueryState<Metadata[]> = {
+    data: [umToken, usdcToken, osmoToken],
+    loading: false,
+  } as ZQueryState<Metadata[]>;
+  const emptySwappableAssets = {
+    data: [] as Metadata[],
+    loading: false,
+  } as ZQueryState<Metadata[]>;
+
+  const umBalance = new BalancesResponse({
+    balanceView: new ValueView({
+      valueView: {
+        case: 'knownAssetId',
+        value: {
+          amount: { hi: 0n, lo: 1n },
+          metadata: umToken,
+        },
+      },
+    }),
+    accountAddress: new AddressView({
+      addressView: {
+        case: 'decoded',
+        value: {
+          index: { account: 0 },
+        },
+      },
+    }),
+  });
+  const usdcBalance = emptyBalanceResponse(usdcToken, 1);
+
+  const balancesResponses = {
+    data: [umBalance, usdcBalance],
+    loading: false,
+  } as ZQueryState<BalancesResponse[]>;
+  const emptyBalancesResponses: ZQueryState<BalancesResponse[]> = {
+    data: [] as BalancesResponse[],
+    loading: false,
+  } as ZQueryState<BalancesResponse[]>;
+
+  let useStore = create<AllSlices>()(initializeStore());
+  beforeEach(() => {
+    // `initializeStore()` initializes the store with all middlewares already
+    // attached, so there's no need to set up a test rig for
+    // `swapBalancesMiddleware` specifically
+    useStore = create<AllSlices>()(initializeStore());
+    vi.restoreAllMocks();
+    vi.stubGlobal('location', { hash: '#swap' });
+  });
+
+  describe('when setting `shared.balancesResponses.data`', () => {
+    describe('when the `from` query param is set', () => {
```

Comment:
I might have over-used one-off `describe`s in this test file 😂 but I do like to use nested describes in the following cases:
1. When there are more than one instance of a given use case, and you don't want to have to repeat the use case in the test names. This also makes for cleaner test output:
    <img width="863" alt="image" src="https://github.com/penumbra-zone/web/assets/1121544/5e476303-29b5-49a5-ae4a-c8bf205df479">
2. When you want to share test setup in a `beforeEach` — e.g., [here](https://github.com/penumbra-zone/web/blob/jessepinho/asset-selector-loading-indicator-web-1386/apps/minifront/src/state/swap/swap-balances-middleware.test.ts#L187-L190), where we want to stub the query parameter for a whole "sub-suite" of tests.
3. When test names would get really long because they're describing a very specific use case — e.g., `it('shows an error message when the array is empty and there is no filter')` could be written with a couple nested `describe`s for `when the array is empty` and `when there is no filter`, especially if we think there may be more tests at any of the nesting levels. (I'm very open to debate on this point, though — that might be overkill.)

Definitely let's discuss at a retro!